storage/innobase/handler/ha_innodb.cc

```diff
@@ -18915,17 +18923,45 @@ wsrep_abort_transaction(
 	      wsrep_thd_query(victim_thd),
 	      victim_trx ? victim_trx->id : 0);
 
-  if (victim_trx)
+  /* if victim does not have transaction in innodb side,
+     we can use server side abort only */
+  if (victim_trx && victim_trx->state != TRX_STATE_NOT_STARTED)
   {
+    /* pick up victim trx id, as the victim may move forward after we release
+       THD lock mutexes */
+    trx_id_t victim_trx_id= victim_trx->id;
+    wsrep_thd_UNLOCK(victim_thd);
+
     lock_mutex_enter();
-    trx_mutex_enter(victim_trx);
-    wsrep_kill_victim(bf_thd, victim_thd, victim_trx, signal);
-    lock_mutex_exit();
-    trx_mutex_exit(victim_trx);
+    /* now we have innodb lock system mutex, but victim trx may have exited by now,
+       find the victim again to see if abort is needed in innodb side anymore */
+    if ((victim_trx= trx_sys.find(bf_trx, victim_trx_id, false)))
```

Comment:
Observed assertions in `trx_sys.find()` because ended up with `victim_trx_id == 0` here, possibly read-only transaction being BF aborted via MDL.